src/assets/scripts/client/trafficGenerator/SpawnPatternModel.js

```diff
@@ -0,0 +1,487 @@
+import _forEach from 'lodash/forEach';
+import _map from 'lodash/map';
+import _isArray from 'lodash/isArray';
+import _isEmpty from 'lodash/isEmpty';
+import _isObject from 'lodash/isObject';
+import _random from 'lodash/random';
+import BaseModel from '../base/BaseModel';
+import RouteModel from '../airport/Route/RouteModel';
+import { bearingToPoint } from '../math/flightMath';
+import { AIRPORT_CONSTANTS } from '../constants/airportConstants';
+import { FLIGHT_CATEGORY } from '../constants/aircraftConstants';
+import { TIME } from '../constants/globalConstants';
+
+/**
+ * Defines a spawn pattern for a specific route within the area
+ *
+ * This can be for departures or arrivals. Provides a starting point and definition
+ * for all spawning aircraft
+ *
+ * @class SpawnPatternModel
+ * @extends BaseModel
+ */
+export default class SpawnPatternModel extends BaseModel {
+    /**
+     * @constructor
+     * @for SpawnPatternModel
+     * @param spawnPatternJson {object}
+     * @param navigationLibrary {NavigationLibrary}
+     */
+    /* istanbul ignore next */
+    constructor(spawnPatternJson, navigationLibrary) {
+        super(spawnPatternJson, navigationLibrary);
+
+        if (!_isObject(spawnPatternJson) || _isEmpty(spawnPatternJson)) {
+            throw new TypeError('Invalid parameter passed to SpawnPatternModel');
+        }
+
+        if (!_isObject(navigationLibrary) || _isEmpty(navigationLibrary)) {
+            throw new TypeError('Invalid NavigationLibrary passed to SpawnPatternModel');
+        }
+
+        /**
+         * Interval id
+         *
+         * Stored here so a specific interval can be associated with a
+         * specfic `SpawnPatternModel` instance. An Interval may be reset
+         * or changed during the life of the app.
+         *
+         * Provides easy access to a specif interval id
+         *
+         * @property scheduleId
+         * @type {number}
+         * @default -1
+         */
+        this.scheduleId = -1;
+
+        /**
+         * One of `FLIGHT_CATEGORY`
+         *
+         * @property category
+         * @type {string}
+         * @default ''
+         */
+        this.category = '';
+
+        /**
+         * Type of arrival or departure pattern
+         *
+         * Could be `random`, `cyclic`. `surge` or `wave`
+         *
+         * @property type
+         * @type {string}
+         * @default ''
+         */
+        this.method = '';
+
+        /**
+         *
+         *
+         * @property origin
+         * @type {string}
+         * @default ''
+         */
+        this.origin = '';
+
+        /**
+         *
+         *
+         * @property destination
+         * @type {string}
+         * @default ''
+         */
+        this.destination = '';
+
+        /**
+         * String representation of a `StandardRoute`
+         *
+         * @property route
+         * @type {string}
+         * @default
+         */
+        this.route = '';
```

Comment:
I mean that *is* the term used to reference it, but it's not like it's commonly used or widely known... I just am becoming more and more wary of naming becoming more generalized as the code gets more and more abstract. The more *specific* we can be in labeling things, while using well-known aviation/ATC terminology, the easier it would be for people with aviation experience to learn the codebase and make bitesize changes. After all, most of the people who have contributed to the project (myself included) were aviators first and programmers second (or worse), so I want to do what we can to cater to those potential future contributors, by keeping everything as understandable as possible by a "layperson".

To a lay person, 'string' (as in "stringing someone along" or "stringing wins together") implies a series of "routes" (like a series of highways), which is right on for the concept here. And to them, 'code' means "da vinci code", "launch codes", "secret code", "activation code", "source code", etc, none of which they are likely to know much about, and don't really lead you in the direction of the purpose of the variable.

What about something like `routeChain`? I don't really like that much better either, but whatever. Also, I don't mean to dwell on this too much, it's just an example of one of the things that have been on my mind since all the refactoring began. :smile: